rfcs/text/0428-extending-types.md

```diff
@@ -0,0 +1,677 @@
+  "type": "object",
+  "title": "Country",
+  "properties": {
+    "https://blockprotocol.org/@alice/property-type/name": {
+      "$ref": "https://blockprotocol.org/@alice/property-type/name/v/1"
+    }
+  },
+  "required": ["https://blockprotocol.org/@alice/property-type/name"]
+}
+```
+
+Entity type `Region` with `Country` as a supertype
+
+```json
+{
+  "kind": "entityType",
+  "$id": "https://blockprotocol.org/@alice/entity-type/region/v/1",
+  "type": "object",
+  "title": "Region",
+  "allOf": [
+    { "$ref": "https://blockprotocol.org/@alice/entity-type/country/v/1" }
+  ],
+  "properties": {
+    "https://blockprotocol.org/@alice/property-type/blurb": {
+      "$ref": "https://blockprotocol.org/@alice/property-type/blurb/v/1"
+    }
+  },
+  "required": ["https://blockprotocol.org/@alice/property-type/blurb"]
+}
+```
+
+The second, cyclic version of the `Country` entity type
+
+```json
+{
+  "kind": "entityType",
+  // Because of versioning, we can not change this version to /v/1 and create a "proper" cycle.
+  "$id": "https://blockprotocol.org/@alice/entity-type/country/v/2",
+  "type": "object",
+  "title": "Region",
+  "allOf": [
+    { "$ref": "https://blockprotocol.org/@alice/entity-type/region/v/1" }
+  ],
+  "properties": {
+    "https://blockprotocol.org/@alice/property-type/name": {
+      "$ref": "https://blockprotocol.org/@alice/property-type/name/v/1"
+    },
+    "https://blockprotocol.org/@alice/property-type/location": {
+      "$ref": "https://blockprotocol.org/@alice/property-type/location/v/1"
+    }
+  },
+  "required": ["https://blockprotocol.org/@alice/property-type/name"]
+}
+```
+
+This sort of type hierarchy should _not_ be accepted within the type extension system, as the circular dependencies lead to hard to reason about types. While the type hierarchy might be completely valid (as it would be in this case), we should safeguard users from making redundant type structures that look like the above.
+
+In this specific contrived example, creating a new entity type based on `Region` instead of a new version of `Country` might even encode semantic meaning better than re-defining `Country`.
+
+## Multiple supertypes - checking compatibility
+
+As described in the [Guide-level explanation](#multiple-supertypes), when extending multiple entity types, they must be able to coexist in a "compatible manner", which means that the entity types' `properties` and `links` comply with the following:
+
+- For each property (base URI on the top level of `properties`) that exists in multiple entity types:
+
+  - all entity types refer to the same versioned URI of the property **or** compatible versions of the property (through the [versioning RFC](./0408-versioning-types.md#determining-type-compatibility) definition of 'compatible')
+  - none of the entity types define the property as an array **or** all define the property as an array with the _exact same_ cardinality constraints.
```

Comment:
Do they have to have the same cardinality constraints? Can a subtype not make it more constrained?

Reply:
Gotcha, I think I'd missed the fact that you're only talking about the supertypes here, don't we also need to apply most of these rules to the subtype?

I'm not sure what you're saying about the same cardinality is true regardless. 

Say we have entity types A, B, C, and C extends A and B

If C has a constraint that x: [0, 1]
and B has a constraint that x: [0, 3]
and A has a constraint that x: [-1, 2]

I think it still works?

C can be used in place of B, and it can be used in place of A.

It's just that A can't be used in place of B (and vice versa) but we don't need that I don't think.

Reply:
Have tried to capture in 12c35de1875725e60b37db07075a1f2e7c0ebe61